ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -733,21 +758,65 @@ private long getTotalBytesRead(StoreFindToken newToken, List<MessageInfo> messag
       Offset logEndOffsetBeforeFind) {
     long bytesRead = 0;
     if (newToken.getType().equals(StoreFindToken.Type.IndexBased)) {
-      bytesRead = log.getDifference(newToken.getOffset(), logAbsoluteZeroOffset);
+      bytesRead = getAbsolutePositionInLogForOffset(newToken.getOffset());
     } else if (newToken.getType().equals(StoreFindToken.Type.JournalBased)) {
       if (messageEntries.size() > 0) {
-        MessageInfo lastMsgInfo = messageEntries.get(messageEntries.size() - 1);
-        Offset offsetInToken = newToken.getOffset();
-        Offset endOffset = new Offset(offsetInToken.getName(), offsetInToken.getOffset() + lastMsgInfo.getSize());
-        bytesRead = log.getDifference(endOffset, logAbsoluteZeroOffset);
+        bytesRead =
+            getAbsolutePositionInLogForOffset(newToken.getOffset()) + messageEntries.get(messageEntries.size() - 1)
+                .getSize();
       } else {
-        bytesRead = log.getDifference(logEndOffsetBeforeFind, logAbsoluteZeroOffset);
+        bytesRead = getAbsolutePositionInLogForOffset(logEndOffsetBeforeFind);
       }
     }
     return bytesRead;
   }
 
   /**
+   * @return the currently capacity used of the log abstraction. Includes "wasted" space at the end of
+   * {@link LogSegment} instances that are not fully filled.
+   */
+  long getLogUsedCapacity() {
+    return getAbsolutePositionInLogForOffset(getCurrentEndOffset());
+  }
+
+  /**
+   * @return the number of valid log segments starting from the first segment.
+   */
+  long getLogSegmentCount() {
+    return getLogSegmentToIndexSegmentMapping().size();
+  }
+
+  /**
+   * Gets the absolute position represented by {@code offset} in the {@link Log}.
+   * @param offset the {@link Offset} whose absolute position is required.
+   * @return he absolute position represented by {@code offset} in the {@link Log}.
+   */
+  long getAbsolutePositionInLogForOffset(Offset offset) {
+    LogSegment logSegment = log.getSegment(offset.getName());
+    if (logSegment == null || offset.getOffset() > logSegment.getEndOffset()) {
+      throw new IllegalArgumentException("Offset is invalid: " + offset);
+    }
+    int numPrecedingLogSegments = getLogSegmentToIndexSegmentMapping().headMap(offset.getName()).size();
+    return numPrecedingLogSegments * log.getSegmentCapacity() + offset.getOffset();
+  }
+
+  /**
+   * @return mapping from log segment names to {@link IndexSegment} instances that refer to them.
+   */
+  private TreeMap<String, List<IndexSegment>> getLogSegmentToIndexSegmentMapping() {
```

Comment:
The index switch needs to be atomic - functions will not take locks to access the `indexes` variable (because that is costly) but it will instead be an `AtomicReference`. That data structure cannot have any transactional dependencies (i.e. change another data structure atomically at the same time). 

Also I don't this operation is very costly and it is not used in the critical path (application PUTs/GETs). The ones that use this are replication, hard delete and metrics reporter and so I think this is quite acceptable for now. If this _really_ a problem, we can tackle it down the line.